File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function(lists) {
    if(lists === null || lists.length === 0) return null;
    lists = lists.filter(a => a !== null && a.length !== 0).sort((a,b) => { 
                         return b.val - a.val});
    
    let res = lists[0];
    for(let i = 1; i < lists.length; i++) {
        res = merge(res, lists[i]);
    }
    return res || null;
};

var merge = function(a, b) {
    let res = b;
    b = b.next;
    let tem = res;
    while(a!== null && b!== null) {
        if(a.val > b.val) {
            tem.next = b;
            b = b.next;
        } else {
            tem.next = a;
            a = a.next;
        }
        tem = tem.next;
    }
    if(a !== null) tem.next = a;
    if(b !== null) tem.next = b;
    return res;
}

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>((a, b) -> a.val - b.val);
        
        for (ListNode list : lists) {
            if (list != null) {
                pq.add(list);
            }
        }

        
        ListNode dummy = new ListNode(0);
        ListNode current = dummy;
        
        // Merge nodes from the priority queue
        while (!pq.isEmpty()) {
            ListNode node = pq.poll();
            current.next = node;   // Link the node to the merged list
            current = current.next; // Move the pointer
            
            // If the polled node has a next node, add it to the queue
            if (node.next != null) {
                pq.add(node.next);
            }
        }
        
        return dummy.next;
    }
}